checks.d/mapreduce.py

```diff
@@ -0,0 +1,449 @@
+
+        return custom_metrics
+
+    def _get_cluster_id(self, rm_address):
+        '''
+        Return the cluster ID, otherwise raise an exception
+        '''
+        info_json = self._rest_request_to_json(rm_address,
+            INFO_PATH,
+            service_name=YARN_SERVICE_CHECK)
+
+        return info_json.get('clusterInfo', {}).get('id')
+
+    def _get_running_app_ids(self, rm_address, **kwargs):
+        '''
+        Return a dictionary of {app_name: tracking_url} for the running MapReduce applications
+        '''
+        metrics_json = self._rest_request_to_json(rm_address,
+            YARN_APPS_PATH,
+            service_name=YARN_SERVICE_CHECK,
+            **kwargs)
+
+        running_apps = {}
+
+        if metrics_json['apps'] is not None:
+            if metrics_json['apps']['app'] is not None:
+
+                for app_json in metrics_json['apps']['app']:
+                    tracking_url = app_json['trackingUrl']
+                    app_name = app_json['name']
+
+                    if tracking_url is not None \
+                            and app_name is not None:
+
+                        running_apps[app_name] = tracking_url
+
+        return running_apps
+
+    def _mapreduce_job_metrics(self, running_apps, cluster_id):
+        '''
+        Get metrics for each MapReduce job.
+        Return a dictionary for each MapReduce job
+        {
+          job_name: {
+            'app_name': app_name,
+            'user_name': user_name,
+            'tracking_url': tracking_url
+        }
+        '''
+        running_jobs = {}
+
+        for app_name, tracking_url in running_apps.iteritems():
+
+            metrics_json = self._rest_request_to_json(tracking_url,
+                MAPREDUCE_JOBS_PATH,
+                service_name=MAPREDUCE_SERVICE_CHECK)
+
+            if metrics_json['jobs'] is not None:
+                if metrics_json['jobs']['job'] is not None:
+
+                    for job_json in metrics_json['jobs']['job']:
+                        job_id = job_json['id']
+                        job_name = job_json['name']
+                        user_name = job_json['user']
+
+                        if job_id is not None \
+                                and job_name is not None \
+                                and user_name is not None:
+
+                            running_jobs[str(job_name)] = {'app_name': str(app_name),
+                                                    'user_name': str(user_name),
+                                                    'tracking_url': self._join_url_dir(tracking_url, MAPREDUCE_JOBS_PATH, job_id)}
+
+                            tags = ['cluster_id:' + str(cluster_id),
+                                    'app_name:' + str(app_name),
+                                    'user_name:' + str(user_name),
+                                    'job_name:' + str(job_name)]
+
+                            self._set_metrics_from_json(tags, job_json, MAPREDUCE_JOB_METRICS)
```

Comment:
Actually there's still an issue here: we're sending the `MAPREDUCE_JOB_METRICS` (all gauges) for each job (ie for each `job_id`), but we only tag by `user_name`/`job_name`, which means that these metrics get overwritten when multiple jobs share the same `job_name`.

This means that once the check has run, for a given `job_name`, the values of these metrics will be the values from the last job in the list that has this `job_name`.